airbyte/_processors/sql/bigquery.py

```diff
@@ -0,0 +1,204 @@
+
+    from airbyte._processors.file.base import FileWriterBase
+    from airbyte.caches.base import CacheBase
+    from airbyte.caches.bigquery import BigQueryCache
+
+
+class BigQueryTypeConverter(SQLTypeConverter):
+    """A class to convert types for BigQuery."""
+
+    @overrides
+    def to_sql_type(
+        self,
+        json_schema_property_def: dict[str, str | dict | list],
+    ) -> sqlalchemy.types.TypeEngine:
+        """Convert a value to a SQL type.
+
+        We first call the parent class method to get the type. Then if the type is VARCHAR or
+        BIGINT, we replace it with respective BigQuery types.
+        """
+        sql_type = super().to_sql_type(json_schema_property_def)
+        # to-do: replace hardcoded return types with some sort of snowflake Variant equivalent
+        if isinstance(sql_type, sqlalchemy.types.VARCHAR):
+            return "String"
+        if isinstance(sql_type, sqlalchemy.types.BIGINT):
+            return "INT64"
+
+        return sql_type.__class__.__name__
+
+
+class BigQuerySqlProcessor(SqlProcessorBase):
+    """A BigQuery implementation of the cache."""
+
+    file_writer_class = JsonlWriter
+    type_converter_class = BigQueryTypeConverter
+    supports_merge_insert = True
+
+    cache: BigQueryCache
+
+    def __init__(self, cache: CacheBase, file_writer: FileWriterBase | None = None) -> None:
+        self._credentials: service_account.Credentials | None = None
+        self._schema_exists: bool | None = None
+        super().__init__(cache, file_writer)
+
+    @final
+    @overrides
+    def _fully_qualified(
+        self,
+        table_name: str,
+    ) -> str:
+        """Return the fully qualified name of the given table."""
+        return f"`{self.cache.schema_name}`.`{table_name!s}`"
+
+    @final
+    @overrides
+    def _quote_identifier(self, identifier: str) -> str:
+        """Return the identifier name as is. BigQuery does not require quoting identifiers"""
+        return f"{identifier}"
+
+    @final
+    @overrides
+    def _get_telemetry_info(self) -> CacheTelemetryInfo:
+        return CacheTelemetryInfo("bigquery")
+
+    def _write_files_to_new_table(
+        self,
+        files: list[Path],
+        stream_name: str,
+        batch_id: str,
+    ) -> str:
+        """Write a file(s) to a new table.
+
+        This is a generic implementation, which can be overridden by subclasses
+        to improve performance.
+        """
+        temp_table_name = self._create_table_for_loading(stream_name, batch_id)
+
+        # Specify the table ID (in the format `project_id.dataset_id.table_id`)
+        table_id = f"{self.cache.project_name}.{self.cache.dataset_name}.{temp_table_name}"
+
+        # Initialize a BigQuery client
+        client = bigquery.Client(credentials=self._get_credentials())
+
+        for file_path in files:
+            with Path.open(file_path, "rb") as source_file:
+                load_job = client.load_table_from_file(  # Make an API request
```

Comment:
I looked through several API options, and the `google-cloud-bigquery` library appears to be the only one which can natively load a jsonl file without first uploading it to GCS. The other libraries were converting the data to CSV or Parquet behind the scenes, which was causing irreconcilable typing issues.